### PR TITLE
BH-1088: register custom taxonomies

### PR DESCRIPTION
**Pull request by nickcernis**
Registers custom taxonomies based on the schema from [BH-1073](https://wpengine.atlassian.net/browse/BH-1073).

### Setup
Update the `atlas_content_modeler_post_types` and `atlas_content_modeler_taxonomies` options with the test data from [BH-1074](https://wpengine.atlassian.net/browse/BH-1074).

With WP-CLI:
```sh
## Create Recipe and Restaurant models.
$ wp option update atlas_content_modeler_post_types '{"recipe":{"show_in_rest":true,"show_in_graphql":true,"singular":"Recipe","plural":"Recipes","slug":"recipe","api_visibility":"public","description":"","fields":{"1624461688230":{"show_in_rest":true,"show_in_graphql":true,"type":"text","id":"1624461688230","position":"0","name":"Name","slug":"name","required":true,"isTitle":true,"inputType":"single","minChars":"","maxChars":""}}},"restaurant":{"show_in_rest":true,"show_in_graphql":true,"singular":"Restaurant","plural":"Restaurants","slug":"restaurant","api_visibility":"private","description":"","fields":{"1624461729538":{"show_in_rest":true,"show_in_graphql":true,"type":"text","id":"1624461729538","position":"0","name":"Name","slug":"name","required":true,"isTitle":true,"inputType":"single","minChars":"","maxChars":""}}}}' --format=json
## Create Ingredient, Cuisine, and Price Range mock taxonomies.
$ wp option update atlas_content_modeler_taxonomies '{"ingredients":{"singular":"Ingredient","plural":"Ingredients","slug":"ingredient","hierarchical":false,"types":["recipe"],"api_visibility":"public","show_in_rest":true,"show_in_graphql":true},"cuisine":{"singular":"Cuisine","plural":"Cuisines","slug":"cuisine","hierarchical":true,"types":["recipe","restaurant"],"api_visibility":"private","show_in_rest":true,"show_in_graphql":true},"price_range":{"singular":"Price Range","plural":"Price Ranges","slug":"price_range","hierarchical":false,"types":["restaurant"],"api_visibility":"private","show_in_rest":false,"show_in_graphql":false}}' --format=json
```

Or update the `wp_options` rows manually with the serialized arrays:

```
/**
 * Serialized for creating options manually in the database.
 */

/** Recipe and Restaurant models. */

/** Create wp_options.atlas_content_modeler_post_types and assign the following value: */
a:2:{s:6:"recipe";a:8:{s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;s:8:"singular";s:6:"Recipe";s:6:"plural";s:7:"Recipes";s:4:"slug";s:6:"recipe";s:14:"api_visibility";s:6:"public";s:11:"description";s:0:"";s:6:"fields";a:1:{i:1624461688230;a:12:{s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;s:4:"type";s:4:"text";s:2:"id";s:13:"1624461688230";s:8:"position";s:1:"0";s:4:"name";s:4:"Name";s:4:"slug";s:4:"name";s:8:"required";b:1;s:7:"isTitle";b:1;s:9:"inputType";s:6:"single";s:8:"minChars";s:0:"";s:8:"maxChars";s:0:"";}}}s:10:"restaurant";a:8:{s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;s:8:"singular";s:10:"Restaurant";s:6:"plural";s:11:"Restaurants";s:4:"slug";s:10:"restaurant";s:14:"api_visibility";s:7:"private";s:11:"description";s:0:"";s:6:"fields";a:1:{i:1624461729538;a:12:{s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;s:4:"type";s:4:"text";s:2:"id";s:13:"1624461729538";s:8:"position";s:1:"0";s:4:"name";s:4:"Name";s:4:"slug";s:4:"name";s:8:"required";b:1;s:7:"isTitle";b:1;s:9:"inputType";s:6:"single";s:8:"minChars";s:0:"";s:8:"maxChars";s:0:"";}}}}

/** Ingredient, Cuisine, and Price Range mock taxonomies. */

/** Create wp_options.atlas_content_modeler_taxonomies and assign the following value: */
a:3:{s:11:"ingredients";a:8:{s:8:"singular";s:10:"Ingredient";s:6:"plural";s:11:"Ingredients";s:4:"slug";s:10:"ingredient";s:12:"hierarchical";b:0;s:5:"types";a:1:{i:0;s:6:"recipe";}s:14:"api_visibility";s:6:"public";s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;}s:7:"cuisine";a:8:{s:8:"singular";s:7:"Cuisine";s:6:"plural";s:8:"Cuisines";s:4:"slug";s:7:"cuisine";s:12:"hierarchical";b:1;s:5:"types";a:2:{i:0;s:6:"recipe";i:1;s:10:"restaurant";}s:14:"api_visibility";s:7:"private";s:12:"show_in_rest";b:1;s:15:"show_in_graphql";b:1;}s:11:"price_range";a:8:{s:8:"singular";s:11:"Price Range";s:6:"plural";s:12:"Price Ranges";s:4:"slug";s:11:"price_range";s:12:"hierarchical";b:0;s:5:"types";a:1:{i:0;s:10:"restaurant";}s:14:"api_visibility";s:7:"private";s:12:"show_in_rest";b:0;s:15:"show_in_graphql";b:0;}}
 ```

### Visual tests
You should see the CPTs in your sidebar, and the taxonomies below them.

Screenshot 2021-06-25 at 14 12 48

You should also see taxonomy meta boxes when creating an entry.

Screenshot 2021-06-25 at 14 12 56

### Manual API tests

Add a restaurant and recipes entry, filling terms for the ingredients, cuisine, and price range taxonomies.

The three taxonomies should have these visibilities and auth requirements in REST and GraphQL:

| Taxonomy  | GraphQL and REST visibility | Requires Auth? |
| ------------- | ------------- | ------------- |
| Ingredients | Visible  | No |  
| Cuisines  | Visible  |  Yes |
| Price Ranges  | Not visible  | n/a |

For example, visiting this REST endpoint should show the ingredients and cuisines taxonomies but not the “price ranges” one:

- http://example.com/wp-json/wp/v2/taxonomies

Visiting this endpoint should show details about the ingredients term with ID 123 (replace with your own term ID):

- http://example.com/wp-json/wp/v2/ingredients/123

Visiting this endpoint without auth should result in `rest_forbidden`:

- http://example.com/wp-json/wp/v2/cuisine/123

A GraphiQL request for this queries should show “ingredients” and “cuisine” in the response but not “price ranges”:

```
			{
				taxonomies {
					nodes {
					  name
					}
				}
			}
```

### CODE REVIEW

**Review by ChrisWiegman — APPROVED** [review]
I'm going to approve this based on its working for me. Are there any applicable e2e tests we can add at this point?

**Review by mindctrl — DISMISSED** [review]

**Comment by nickcernis** [comment]
> Are there any applicable e2e tests we can add at this point?

Good question! I started to write some and realised they'd be easier to add when we have the “add taxonomy” UI in place (in progress in #183). The e2e test can then add the taxonomy and confirm that the taxonomy appears under the relevant CPT.

File: atlas-content-modeler.php [hunk 1]
```diff
@@ -38,6 +38,8 @@ function atlas_content_modeler_loader(): void {
 		'shared-assets/wp_scripts/shared_assets.php',
 		'settings/settings-callbacks.php',
 		'content-registration/custom-post-types-registration.php',
+		'content-registration/register-taxonomies.php',
+		'content-registration/class-wpe-rest-terms-controller.php',
```
Comment (apmatthews) [inline atlas-content-modeler.php:42]: I think we could eventually extract a lot of our REST code to the `rest-api` folder as part of a larger reorganization, but I agree it makes the most sense to put it where you did for now.

File: includes/content-registration/class-wpe-rest-terms-controller.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace WPE\AtlasContentModeler\ContentRegistration\Taxonomies;
+
+/**
+ * Class REST_Terms_Controller
+ */
+final class REST_Terms_Controller extends \WP_REST_Terms_Controller {
```
Comment (apmatthews) [inline includes/content-registration/class-wpe-rest-terms-controller.php:20]: I know the other REST controller followed this same pattern, but I _think_ WordPress standards like the file name and class name to match. So `WPE_REST_Terms_Controller` or `class-rest-terms-controller.php`.

File: includes/content-registration/register-taxonomies.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		'rest_controller_class' => __NAMESPACE__ . '\REST_Terms_Controller',
+	);
+}
+
+/**
+ * Gets custom ACM taxonomies saved by the user.
+ *
+ * @since 0.6.0
+ */
+function get_taxonomies(): array {
```
Comment (apmatthews) [inline includes/content-registration/register-taxonomies.php:135]: When reading the code, I got slightly confused between usage of this `get_taxonomies()` and the WP version. I like the simplicity, but wonder if `get_acm_taxonomies()` would be better?

**Review by apmatthews — COMMENTED** [review]
Looks great and seems to be working as expected!

I wonder if we should be concerned about users confusing the scope of the API Visibility setting? It does currently prevent you from reading the terms inside of a private taxonomy. However, you can still see that a private taxonomy exists (/wp/v2/taxonomies) or see what private taxonomy term IDs are assigned to a particular post. For example, visiting `/wp/v2/recipes` will show you the IDs of cuisines assigned to the recipe. I'm having trouble imagining what a user would expect when choosing to make a taxonomy `private`.

File: includes/content-registration/class-wpe-rest-terms-controller.php (hunk 2 above)
Comment (nickcernis) [reply includes/content-registration/class-wpe-rest-terms-controller.php:20]: You're right, I'll make that change.

It looks like we're ignoring file naming conventions at the moment, which is why they're not flagged by phpcs. (Not for any strong reason as far as I know — I think we just copied a phpcs config from another project.)

https://github.com/wpengine/atlas-content-modeler/blob/3e51e165bad9be72d6499ea6e3ab0732dc276c55/.phpcs.xml.dist#L43-L44

Comment (nickcernis) [reply includes/content-registration/class-wpe-rest-terms-controller.php]: Changed in fe2ff1d. Thanks again for the suggestion!

File: includes/content-registration/register-taxonomies.php (hunk 3 above)
Comment (nickcernis) [reply includes/content-registration/register-taxonomies.php]: Good idea, fixed in 6eef15b.

**Comment by nickcernis** [comment]
Updated to address feedback but not ready for re-review yet.

I'm exploring whether or not the custom controller should extend `WP_REST_Taxonomies_Controller` instead of `WP_REST_Terms_Controller` to address the issue @apmatthews raised where terms are protected in REST responses if API Visibility is “private” but taxonomies themselves are not. https://github.com/wpengine/atlas-content-modeler/pull/180#pullrequestreview-695111288.

**Comment by nickcernis** [comment]
I believe that https://github.com/wpengine/atlas-content-modeler/pull/180#pullrequestreview-695111288 is addressed in https://github.com/wpengine/atlas-content-modeler/pull/180/commits/0e9a160b9c6bd0093440b6f1d19c2ba35fdd0de7.

This is ready for another look.

**Review by mindctrl — APPROVED** [review]

**Review by apmatthews — APPROVED** [review]

**Pull request merged**